lib/puppet/pops/types/p_object_type.rb

```diff
@@ -0,0 +1,539 @@
+    attr_reader :name
+
+    # @return [PAnyType] the type of this member
+    # @api public
+    attr_reader :type
+
+    # @return [Hash{PType => Hash}] the annotations or `nil`
+    # @api public
+    attr_reader :annotations
+
+    # @param name [String] The name of the member
+    # @param [Hash{String=>Object}] i12n_hash
+    # @option i12n_hash [PAnyType] 'type' The member type (required)
+    # @option i12n_hash [Boolean] 'override' `true` if this feature must override an inherited feature. Default is `false`.
+    # @option i12n_hash [Boolean] 'final' `true` if this feature cannot be overridden. Default is `false`.
+    # @option i12n_hash [Hash{PType => Hash}] 'annotations' Annotations hash. Default is `nil`.
+    # @api public
+    def initialize(name, i12n_hash)
+      @name = name
+      @type = i12n_hash[KEY_TYPE]
+      @override = i12n_hash[KEY_OVERRIDE]
+      @override = false if @override.nil?
+      @final = i12n_hash[KEY_FINAL]
+      @final = false if @final.nil?
+      @annotations = i12n_hash[KEY_ANNOTATIONS]
+      @annotations.freeze unless @annotations.nil?
+    end
+
+    # Delegates to the contained type
+    # @param visitor [TypeAcceptor] the visitor
+    # @param guard [RecursionGuard] guard against recursion. Only used by internal calls
+    # @api public
+    def accept(visitor, guard)
+      @type.accept(visitor, guard)
+      @annotations.each_key { |key| key.accept(visitor, guard) } unless @annotations.nil?
+    end
+
+    # Checks if the this _member_ overrides an inherited member, and if so, that this member is declared with override = true and that
+    # the inherited member accepts to be overridden by this member.
+    #
+    # @param parent_members [Hash{String=>PAnnotatedMember}] the hash of inherited members
+    # @return [PAnnotatedMember] this instance
+    # @raises [Puppet::ParseError] if the assertion fails
+    # @api private
+    def assert_override(parent_members)
+      parent_member = parent_members[@name]
+      if parent_member.nil?
+        raise Puppet::ParseError, "expected #{label} to override an inherited #{feature_type}" if @override
+        self
+      else
+        parent_member.assert_can_be_overridden(self)
+      end
+    end
+
+    # Checks if the given _member_ can override this member.
+    #
+    # @param member [PAnnotatedMember] the overriding member
+    # @return [PAnnotatedMember] its argument
+    # @raises [Puppet::ParseError] if the assertion fails
+    # @api private
+    def assert_can_be_overridden(member)
+      raise Puppet::ParseError, "#{member.label} overrides inherited #{feature_type}" unless self.class == member.class
+      raise Puppet::ParseError, "attempt to override final #{label}" if @final
+      raise Puppet::ParseError, "#{label} attempts to override without having override defined" unless member.override?
+      raise Puppet::ParseError, "attempt to override #{label} with a type that does not match" unless @type.assignable?(member.type)
```

Comment:
The problem is that all attributes and functions live in an Object. The Object is in itself anonymous. It may be aliased, and the parent may be such an alias, but there's no guarantee that it is. For some errors where you've had the same objection, it's not even certain that there is a parent.

We could change this of course, and not use a type alias as the mechanism to name an otherwise anonymous declaration. That way, all Objects would be guaranteed to have a name.